kubespawner/spawner.py

```diff
@@ -244,14 +292,55 @@ def __init__(self, *args, **kwargs):
         """
     )
 
+    enable_user_namespaces = Bool(
+        False,
+        config=True,
+        help="""
+        Cause each user to be spawned into an individual namespace.
+
+        This comes with some caveats.  The Hub must run with significantly
+        more privilege (must have ClusterRoles analogous to its usual Roles)
+        and can therefore do heinous things to the entire cluster.
+
+        It will also make the Reflectors aware of pods and events across
+        all namespaces.  This will have performance implications, although
+        using labels to restrict resource selection helps somewhat.
+
+        If you use this, consider cleaning up the user namespace in your
+        post_stop_hook.
```

Comment:
Should cleaning up the user namespace be something we do by default?

Reply:
I don't think so.  One of the early comments here wanted to do this with pre-created namespaces.  The existing code here works fine with that, because if the namespace is extant, you get a 409 back and keep going.

But if you have pre-created namespaces, then presumably you want them to stick around in the absence of a user's running pod.

There are also some subtleties to the teardown: I can imagine a case in which the user wants to stop their running pod but have the namespace sticking around doing batch work (we're currently using Argo Workflows for something like this).  Thus I think it's better to tell the user where they can clean it up (or I could add another configurable).

Reply:
Makes sense. thanks! I think it will be worth making it a bool configurable rather than requiring generic logic in a hook, but that can be its own PR.

Reply:
I agree with both those sentiments.  I'd like to land this one, then I'll make that a configurable (Bool | Callable, maybe?)  or maybe just Bool and if you want to do something fancy, make it False and add the hook.

Then (or maybe first) I'll do a PR to blacken it and to add blacken-pre-commit.